File: quoteparser/src/main/java/org/zenix/testing/quoteparser/parsers/ParserYear.java
package org.zenix.testing.quoteparser.parsers;


import org.zenix.testing.quoteparser.models.Quote;

public class ParserYear implements ParsableQuote {
    // Note, currently doesnt work with Æ, Ø "non ascii" characters
    public static final String QUOTE_PATTERN = "^.+\\n?\\s*-\\s*\\w+\\s\\d{4}$";

    @Override
    public Quote parse(String rawString) {
        if (rawString == null) {
            throw new NullPointerException();
        }
        String[] split = rawString.split("-");
        if (isCorrectFormat(rawString) && split.length != 2) {
            throw new QuoteParsingException("String cannot be parsed! It's in a wrong format.   Expected format: " + getExpectedFormatString());
        }

        String quoteText = split[0].trim();
        String rest = split[1].trim();

        String textWithoutQuotes = ParserUtility.trimQuotesIfExists(quoteText);

        // Get the 4 last digits, the year. The the rest should be the author.
        String year = rest
                .substring(rest.length() - 4, rest.length())
                .trim();
        String author = rest
                .substring(0, rest.length() - 4)
                .trim();

        return new Quote(author, textWithoutQuotes, year);
    }

    @Override
    public boolean isCorrectFormat(String rawString) {
        if (rawString == null) {
            return false;
        }

        return rawString.matches(QUOTE_PATTERN);
    }

    @Override
    public String getExpectedFormatString() {
        return "<quotetext> <hyphen> <author> <single space> <year>";
    }
}

File: quoteparser/src/main/kotlin/org/zenix/testing/quoteparser/parsers/ParserYear.kt
package org.zenix.testing.quoteparser.parsers

import org.zenix.testing.quoteparser.models.Quote


class ParserYear : ParsableQuote {
    private val quotePattern = ("^.+\\n?\\s*-\\s*\\w+\\s\\d{4}$").toRegex()

    override fun parse(rawString: String): Quote {
        val trimmedString = rawString.filter { c -> c != '\n' }
        val length = trimmedString.length
        val lastHyphenIndex = trimmedString.indexOfLast { c -> c == '-' }


        val year = trimmedString.substring(length - 4, length)
                .trim()
        val author = trimmedString.substring(lastHyphenIndex, length - 4)
                .filter { c -> c != '-'}
                .trim()
        val text = trimmedString.substring(0, lastHyphenIndex)
                .filter { c -> c != '"' }
                .trim()

        return Quote(author, text, year)
    }

    override fun isCorrectFormat(rawString: String): Boolean {
        return rawString.matches(quotePattern)
    }

    override fun getExpectedFormatString(): String {
        return "<quotetext> <hyphen> <author> <single space> <year>"
    }
}